### PR TITLE
Performance improvements and `neuron_parallel_compile` and gradient checkpointing fixes

### PR DESCRIPTION
**Pull request by michaelbenayoun**
# What does this PR do?

- Fixes `neuron_parallel_compile` which fails as soon as it sees a character in the log that it cannot decode. Related issue: https://github.com/aws-neuron/aws-neuron-sdk/issues/886
- Fixes gradient checkpointing. But a lot that can be skipped for now and deferred to another PR is marked as "TODO".
- Changes the `xm.mark_step()` position in the training loop. Before it was after the `optimizer.step()` which is not ideal at all because it will produce more graphs than the current proposed solution. The proposed solution, on top of making more sense, is also more aligned with what the Neuron compiler expects, providing good speed-up in compilation time.
- Loss log and model saving are now step closures. The benefit of doing that compared to what was done before is that these functions will be executed sequentially at the next `xm.mark_step()` (so the next batch being fetched). While it could trigger compilation (and thus more graphs) before, now it will wait for the actual tensors to be materialized (via the `xm.mark_step()` to be run, avoiding the creations of new graphs.

### CODE REVIEW

**Comment by dacorvo** [comment]
Is this a transient error ?
```
tests/distributed/test_model_parallelization.py::TestModelParallelization::test_parallel_model_matches_original_model_from_pretrained_with_parallel_embeddings_and_sequence_parallel[LlamaForQuestionAnswering-dp=4,tp=pp=2] - torch.fx.proxy.TraceError: symbolically traced variables cannot be used as inputs to control flow
```

File: optimum/neuron/accelerate/utils/misc.py [hunk 1]
```diff
@@ -191,6 +195,41 @@ def tie_parameters(model: Union["torch.nn.Module", "NxDPPModel"], tied_parameter
...
+        self._set_gradient_checkpointing(enable=True, gradient_checkpointing_func=gradient_checkpointing_func)
+    else:
+        self.apply(functools.partial(self._set_gradient_checkpointing, value=True))
+        logger.warning(
+            "You are using an old version of the checkpointing format that is deprecated (We will also silently ignore `gradient_checkpointing_kwargs` in case you passed it)."
+            "Please update to the new format on your modeling file. To use the new format, you need to completely remove the definition of the method `_set_gradient_checkpointing` in your model."
+        )
+
+    if getattr(self, "_hf_peft_config_loaded", False):
+        # When using PEFT + gradient checkpointing + Trainer we need to make sure the input has requires_grad=True
```
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply optimum/neuron/accelerate/utils/misc.py:226]: Nope, but I started working on it today. It is just a copy / paste from `transformers`.

**Comment by michaelbenayoun** [comment]
@dacorvo no it's broken. I have added that to "CLASSES_TO_IGNORE" in the tests, let me check why it runs it...

**Comment by michaelbenayoun** [comment]
Filtered now!

**Review by dacorvo — APPROVED** [review]
I trust you on this, because it is a bit hard to decipher without grasping the whole context of // training.

**Comment by dacorvo** [comment]
> I trust you on this, because it is a bit hard to decipher without grasping the whole context of // training.

But the explanations in the description are really clear: it is just that it is hard to follow how it is translated in the code.
